### DISCUSSION TITLE
Bugfix: Remove Make ARGS from GSL installation

### DISCUSSION THREAD

**Discussion by HathewayWill** [dtcenter/METplus#2438] [Installation] [answered]
## Describe the Problem ##
When using Compile_MET:
https://raw.githubusercontent.com/dtcenter/MET/main_v11.1/internal/scripts/installation/compile_MET_all.sh

MAKE_ARGS causes gsl to fail.

### Expected Behavior ###
MAKE_ARGS shouldn't be required for GSL

### Environment ###
Describe your runtime environment:
*1. Machine: Desktop 
*2. OS: Ubuntu 22.04.2
*3. GNU

### To Reproduce ###
Using the compile code set MAKE_ARGS= - j ## of cores

	chmod 775 compile_MET_all.sh

	time ./compile_MET_all.sh 2>&1 | tee compile_MET_all.log
	
	This will cause an error with GSL.

FIX:

  run_cmd "./configure --prefix=${LIB_DIR} > gsl.configure.log 2>&1"
  run_cmd "make ${MAKE_ARGS} > gsl.make.log 2>&1"
  run_cmd "make ${MAKE_ARGS} install > gsl.make_install.log 2>&1"

to

  run_cmd "./configure --prefix=${LIB_DIR} > gsl.configure.log 2>&1"
  run_cmd "make > gsl.make.log 2>&1"
  run_cmd "make  install > gsl.make_install.log 2>&1"

I have tested it without using make_args and MET installs properly.

Seems to be the issue.  I also vaguely remember a PR for this issue once before.

**Comment by jprestop** [comment]
Hi @HathewayWill.  The link in the description does not work for me.  Is the following what you intended to link to?
https://raw.githubusercontent.com/dtcenter/MET/main_v11.0/internal/scripts/installation/compile_MET_all.sh

If so, I see that GSL is not using MAKE_ARGS:
```
  echo
  echo "Compiling GSL_${vrs} at `date`"
  mkdir -p ${LIB_DIR}/gsl
  cd ${LIB_DIR}/gsl
  rm -rf gsl*
  tar -xf ${TAR_DIR}/gsl-${vrs}.tar.gz
  cd gsl*
  echo "cd `pwd`"
  echo "./configure --prefix=${LIB_DIR} > configure.log 2>&1"
  ./configure --prefix=${LIB_DIR} > configure.log 2>&1
  ret=$?
  if [ $ret != 0 ]; then
    echo "configure returned with non-zero ($ret) status"
    exit 1
  fi
  echo "make > make.log 2>&1"
  make > make.log 2>&1
  ret=$?
  if [ $ret != 0 ]; then
    echo "make returned with non-zero ($ret) status"
    exit 1
  fi
  echo "make install > make_install.log 2>&1"
  make install > make_install.log 2>&1
  ret=$?
  if [ $? != 0 ]; then
    echo "make install returned with non-zero ($ret) status"
    exit 1
  fi
fi
```

If you intended to link to the version in develop:
https://raw.githubusercontent.com/dtcenter/MET/develop/internal/scripts/installation/compile_MET_all.sh

we do use MAKE_ARGS:
```
# Compile GSL
if [ $COMPILE_GSL -eq 1 ]; then

  if [ ${COMPILER_FAMILY} = "pgi" ]; then
    vrs="1.11";
  else
    vrs="2.7.1";
  fi

  echo
  echo "Compiling GSL_${vrs} at `date`"
  mkdir -p ${LIB_DIR}/gsl
  rm -rf ${LIB_DIR}/gsl/gsl*
  tar -xf ${TAR_DIR}/gsl-${vrs}.tar.gz -C ${LIB_DIR}/gsl
  cd ${LIB_DIR}/gsl/gsl*
  echo "cd `pwd`"
  run_cmd "./configure --prefix=${LIB_DIR} > gsl.configure.log 2>&1"
  run_cmd "make ${MAKE_ARGS} > gsl.make.log 2>&1"
  run_cmd "make ${MAKE_ARGS} install > gsl.make_install.log 2>&1"
fi
```

This code runs successfully in our GitHub Actions automated testing using a value of "-j 5", an upgraded version of GSL, and the GNU compilers.

Have you tried adding a value for the "-j" option?  Please try using "5" and let us know how it goes.

**Comment by HathewayWill** [comment]
# Compile GSL
if [ $COMPILE_GSL -eq 1 ]; then

  if [ ${COMPILER_FAMILY} = "pgi" ]; then
    vrs="1.11";
  else
    vrs="2.1";
  fi

  echo
  echo "Compiling GSL_${vrs} at `date`"
  mkdir -p ${LIB_DIR}/gsl
  rm -rf ${LIB_DIR}/gsl/gsl*
  tar -xf ${TAR_DIR}/gsl-${vrs}.tar.gz -C ${LIB_DIR}/gsl
  cd ${LIB_DIR}/gsl/gsl*
  echo "cd `pwd`"
  run_cmd "./configure --prefix=${LIB_DIR} > gsl.configure.log 2>&1"
  run_cmd "make ${MAKE_ARGS} > gsl.make.log 2>&1"
  run_cmd "make ${MAKE_ARGS} install > gsl.make_install.log 2>&1"
fi

https://raw.githubusercontent.com/dtcenter/MET/main_v11.1/internal/scripts/installation/compile_MET_all.sh

**Reply by jprestop** [reply]
@HathewayWill Thanks for letting us know which version of the script you are using.  Have you tried adding a value for the "-j" option? If so, can you please send your config.log file for MET along with the screen output of the compile_MET_all.sh script?

**Reply by HathewayWill** [reply]
Good afternoon @jprestop 

So I have tried just -j and it errors out at HDF5.

[minus_j_only.zip](https://github.com/dtcenter/METplus/files/13444736/minus_j_only.zip)


When -j 16 (half my cpu) is used it fails at gsl
[minus_j_halfcpu.zip](https://github.com/dtcenter/METplus/files/13444737/minus_j_halfcpu.zip)

It works as expected when these changes are implemented (see modified compile v11.1)



[pass.zip](https://github.com/dtcenter/METplus/files/13444900/pass.zip)

**Reply by jprestop** [reply]
Thank you for this additional information.  Could you please try a value of "-j 5"?

**Reply by HathewayWill** [reply]
Good morning and happy thanksgiving @jprestop 

Here is the -j 5 files

looks like that worked not sure why though.


I also tried -j 6 and -j 7.  -j 6 with install -j 7 fails


[additional tests.zip](https://github.com/dtcenter/METplus/files/13450898/additional.tests.zip)

**Reply by jprestop** [reply]
Thanks @HathewayWill for this helpful information.  I'll talk more with our developers tomorrow at a meeting about how best to handle this situation and will follow up once I know how we will proceed.

**Reply by HathewayWill** [reply]
@jprestop 

sounds good

**Reply by jprestop** [reply, marked as answer]
Hi Will.  We discussed this situation and decided to proceed with leaving MAKE_ARGS in place and recommending a value of "-j 5" as now noted in our [install_met_env.generic](https://github.com/dtcenter/MET/blob/main_v11.1/internal/scripts/installation/config/install_met_env.generic) config file.   Please let us know if you have any further questions or concerns, otherwise please select "Mark as answer" and I will go ahead and lock this Discussion.  Thanks!

**Reply by HathewayWill** [reply]
@jprestop sounds good. I'll adjust my codes to reflect this. 

What's the reason behind why it can't use more then 5?

**Reply by jprestop** [reply]
Hi @HathewayWill.  Unfortunately, I cannot answer your question.  Perhaps the developers of GSL would be able to provide an answer.

**Reply by HathewayWill** [reply]
@jprestop 

Okay we can mark it as closed.

Will the new 12.0 be using the newer gsl library?

**Reply by jprestop** [reply]
@HathewayWill the 12.0.0 release will include the following changes:

- Changing development of MET code with Grib2C from 1.6.0 to 1.6.4
- Changing development of MET code with Jasper from 1.900.1 to 2.0.25
- Changing development of MET code with GSL from 2.1 to 2.7
- Changing development of MET code with BUFRLIB from 11.3.0 to 11.6.0
- Adding development of MET code with Proj 7.1.0
- Adding development of MET code with Atlas 0.30.0
- Adding development of MET code with EcKit 1.20.2
- and ensure that MET and its dependencies compile with the new Intel oneAPI Compilers.

However, users aren't restricted to those versions and can use other versions of the software dependencies. Those listed above are what we will provide in our tar_files package.
